### PR TITLE
Rsync function to the delivery package

### PR DESCRIPTION
**Pull request by henningonsbring**
## Description
This PR adds the command `cg deliver rsync -t <ticket_id>`, which will rsync a folder in the customers inbox on hasta named as the given ticket-id to the customers inbox on caesar. This command also support delivery of covid results to a dedicated covid folder via the `--covid` flag.

### How to prepare for test
- [x] ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] paxa the environment: `paxa`
- [x] install on stage (example for hasta):
`bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-cg-stage.sh rsync_function`

### How to test
- [x] do `cg deliver rsync --dry-run -t 795106`
- [x] do `cg deliver rsync --dry-run --covid -t 795106`

### Expected test outcome
- [x] check that the correct command is printed: `rsync -rvL --progress /home/proj/stage/customers/cust000/inbox/643594/ caesar.scilifelab.se:/home/cust000/inbox/643594/`
- [x] Take a screenshot and attach or copy/paste the output.
- [x] check that the correct commands are printed:
`rsync -rvL --progress /home/proj/stage/customers/cust000/inbox/643594/ caesar.scilifelab.se:/home/cust000/inbox/643594/`
and
`rsync -rvL --progress /home/proj/production/mutant/cases/steadyworm/results/sars-cov-2_795106_results_*.csv caesar.scilifelab.se:/home/cust032/inbox/wwLab_automatisk_hamtning/`
- [x] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] code approved by
- [x] tests executed by @henningonsbring 
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan
- [ ] Document in ...
- [ ] Deploy this branch
- [ ] Inform to ...

### CODE REVIEW

**Comment by henningonsbring** [comment]
```
cg deliver rsync --dry-run -t 795106
Dry-run activated, skipping running the command: rsync -rvL --progress /home/proj/stage/customers/cust032/inbox/795106/ caesar.scilifelab.se:/home/cust032/inbox/795106/
```

**Comment by henningonsbring** [comment]
```
cg deliver rsync --dry-run --covid -t 795106
Dry-run activated, skipping running the command: rsync -rvL --progress /home/proj/stage/customers/cust032/inbox/795106/ caesar.scilifelab.se:/home/cust032/inbox/795106/
Dry-run activated, will not send COVID results via the command: rsync -rvL --progress /home/proj/production/mutant/cases/steadyworm/results/sars-cov-2_795106_results_*.csv caesar.scilifelab.se:/home/cust032/inbox/wwLab_automatisk_hamtning/
```

File: cg/cli/deliver/base.py [hunk 1]
```diff
@@ -22,6 +24,7 @@ def deliver(context):
     LOG.info("Running CG deliver")
     context.obj["status_db"] = Store(context.obj["database"])
     context.obj["housekeeper_api"] = HousekeeperAPI(context.obj)
+    context.obj["rsync_api"] = RsyncAPI(context.obj)
```
Comment (Mropat) [inline cg/cli/deliver/base.py:27]: do we need this ?

Comment (henningonsbring) [reply cg/cli/deliver/base.py:27]: I think you are right, this line might not be needed. I will test without it and then remove if not needed

Comment (henningonsbring) [reply cg/cli/deliver/base.py]: I removed that line now and everything still works as it should when I test on stage

File: cg/meta/rsync.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        return cmd
+
+    def generate_covid_rsync_command(self, ticket_id: int) -> str:
+        case_id = self.get_case_id(ticket_id=ticket_id)
+        source_path = (
+            "/home/proj/production/mutant/cases/%s/results/sars-cov-2_%s_results_*.csv "
+            % (case_id, ticket_id)
+        )
+        customer_id = self.get_customer_id(ticket_id=ticket_id)
+        destination_path = "caesar.scilifelab.se:/home/%s/inbox/wwLab_automatisk_hamtning/" % (
```
Comment (Mropat) [inline cg/meta/rsync.py:40]: maybe this should be moved to servers repo? cg is public and although this is not exposing anything, we still dont want to reveal how we sync cust data

Comment (Mropat) [reply cg/meta/rsync.py:40]: also if oved to servers repo, you can put a test destination for staging to test the rsync with

File: cg/meta/rsync.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            customer_id
+        )
+        rsync_string = "rsync -rvL --progress "
```
Comment (Mropat) [inline cg/meta/rsync.py:43]: should this be a list?
also, why --progress?

File: cg/cli/deliver/base.py [hunk 4]
```diff
@@ -183,5 +187,39 @@ def deliver_old_analysis(context, case_id: str, ticket_id: int, delivery_type: s
...
+    """
+    inbox = context.obj.get("delivery_path")
+    status_db = context.obj["status_db"]
+    rsync_api = RsyncAPI(store=status_db)
+    cmd = rsync_api.generate_rsync_command(ticket_id=ticket_id, base_path=str(inbox))
+    if dry_run:
+        LOG.info("Dry-run activated, skipping running the command: %s", cmd)
+    else:
+        LOG.info("Command ran: {}".format(cmd))
+        proc = subprocess.Popen(cmd.split())
```
Comment (Mropat) [inline cg/cli/deliver/base.py:208]: I suggest to use Process from cg.utils for this instead!

File: cg/meta/rsync.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+    def get_customer_id(self, ticket_id: int) -> str:
+        case_obj = self.get_case_object(ticket_id=ticket_id)
+        customer_id = case_obj.customer.internal_id
+        return customer_id
+
+    def generate_rsync_command(self, ticket_id: int, base_path: str) -> str:
+        customer_id = self.get_customer_id(ticket_id=ticket_id)
+        destination_path = "caesar.scilifelab.se:/home/%s/inbox/%s/" % (customer_id, ticket_id)
+        source_path = base_path + "/" + customer_id + "/inbox/" + str(ticket_id) + "/ "
```
Comment (Mropat) [inline cg/meta/rsync.py:28]: Suggest use pathlib.Path for path concatenation! gets more readable and harder to miss symbols and such. You can also verify if path youre trying to rsync exists

File: cg/meta/rsync.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        rsync_string = "rsync -rvL --progress "
+        cmd = rsync_string + source_path + destination_path
+        return cmd
+
+    def generate_covid_rsync_command(self, ticket_id: int) -> str:
+        case_id = self.get_case_id(ticket_id=ticket_id)
+        source_path = (
+            "/home/proj/production/mutant/cases/%s/results/sars-cov-2_%s_results_*.csv "
```
Comment (Mropat) [inline cg/meta/rsync.py:36]: you can also use servers config to get this path - so we dont reveal our dir tree on public repo!

**Comment by henrikstranneheim** [comment]
@henningonsbring Would your PR be affected by this: https://github.com/Clinical-Genomics/IT-issues/issues/208?

**Comment by henningonsbring** [comment]
> @henningonsbring Would your PR be affected by this: [Clinical-Genomics/IT-issues#208](https://github.com/Clinical-Genomics/IT-issues/issues/208)?

I have given a detailed answer in https://github.com/Clinical-Genomics/IT-issues/issues/208

To me the command seems fine, even though I agree with Maria that the path should be hidden from the public repo

File: cg/cli/deliver/base.py (hunk 4 above)
Comment (moonso) [reply cg/cli/deliver/base.py:208]: Yes please use that

**Comment by henningonsbring** [comment]
```
cg deliver rsync --dry-run -t 795106
Use base call ['rsync']
Running command rsync -rvL /home/proj/stage/customers/cust032/inbox/795106/ caesar.scilifelab.se:/home/cust032/inbox/795106/
Dry run: process call will not be executed!!
```

```
cg deliver rsync --dry-run --covid -t 795106
Use base call ['rsync']
Running command rsync -rvL /home/proj/stage/customers/cust032/inbox/795106/ caesar.scilifelab.se:/home/cust032/inbox/795106/
Dry run: process call will not be executed!!
Running command rsync -rvL /home/proj/stage/mutant/cases/steadyworm/results/sars-cov-2_795106_results_*.csv caesar.scilifelab.se:/home/cust032/inbox/wwLab_automatisk_hamtning/
Dry run: process call will not be executed!!
```

**Comment by Mropat** [comment]
```
[0|0|253] 3d [maryia.ropart@hasta:~] [S_main] 9s $ cg deliver rsync --dry-run 795106
Running CG deliver
Instantiating housekeeper api
Initialised HousekeeperAPI: <sqlalchemy.orm.scoping.scoped_session object at 0x7fe01494dda0>
Instantiating trailblazer api
Instantiating status db
Instantiating lims api
Instantiating hermes api
Use base call ['/home/proj/bin/conda/envs/S_hermes/bin/hermes']
Instantiating scout api
Use base call ['/home/proj/bin/conda/envs/S_scout/bin/scout', '--config', '/home/proj/stage/servers/config/hasta.scilifelab.se/scout-stage.yaml']
Instantiating vogue api
Use base call ['/home/proj/bin/conda/envs/S_vogue/bin/vogue', '--config', '/home/proj/stage/servers/config/hasta.scilifelab.se/vogue-stage.yaml']
Instantiating crunchy api
Use base call ['sbatch']
Instantiating madeline api
Use base call ['/home/proj/stage/servers/resources/hasta.scilifelab.se/madeline2']
Instantiating mutacc_auto api
Instantiating genotype api
Use base call ['/home/proj/bin/conda/envs/S_genotype/bin/genotype', '--config', '/home/proj/stage/servers/config/hasta.scilifelab.se/genotype-stage.yaml']
Instantiating chanjo api
Use base call ['/home/proj/bin/conda/envs/S_chanjo/bin/chanjo', '--config', '/home/proj/stage/servers/config/hasta.scilifelab.se/chanjo-stage.yaml']
Use base call ['rsync']
Running command rsync -rvL /home/proj/stage/customers/cust032/inbox/795106 caesar.scilifelab.se:/home/cust032/inbox/795106/
Dry run: process call will not be executed!!
```

**Comment by Mropat** [comment]
```
[0|0|252] 3d [maryia.ropart@hasta:~] [S_main] 24s $ cg deliver rsync --dry-run 137683
Running CG deliver
Instantiating housekeeper api
Initialised HousekeeperAPI: <sqlalchemy.orm.scoping.scoped_session object at 0x7f7890f62cf8>
Instantiating trailblazer api
Instantiating status db
Instantiating lims api
Instantiating hermes api
Use base call ['/home/proj/bin/conda/envs/S_hermes/bin/hermes']
Instantiating scout api
Use base call ['/home/proj/bin/conda/envs/S_scout/bin/scout', '--config', '/home/proj/stage/servers/config/hasta.scilifelab.se/scout-stage.yaml']
Instantiating vogue api
Use base call ['/home/proj/bin/conda/envs/S_vogue/bin/vogue', '--config', '/home/proj/stage/servers/config/hasta.scilifelab.se/vogue-stage.yaml']
Instantiating crunchy api
Use base call ['sbatch']
Instantiating madeline api
Use base call ['/home/proj/stage/servers/resources/hasta.scilifelab.se/madeline2']
Instantiating mutacc_auto api
Instantiating genotype api
Use base call ['/home/proj/bin/conda/envs/S_genotype/bin/genotype', '--config', '/home/proj/stage/servers/config/hasta.scilifelab.se/genotype-stage.yaml']
Instantiating chanjo api
Use base call ['/home/proj/bin/conda/envs/S_chanjo/bin/chanjo', '--config', '/home/proj/stage/servers/config/hasta.scilifelab.se/chanjo-stage.yaml']

Use base call ['rsync']
Running command rsync -rvL /home/proj/stage/customers/cust124/inbox/137683 caesar.scilifelab.se:/home/cust124/inbox/137683/
Dry run: process call will not be executed!!
Delivering report for SARS-COV-2 analysis
Running command rsync -rvL /home/proj/stage/mutant/cases/toplion/results/sars-cov-2_137683_results_*.csv caesar.scilifelab.se:/home/cust124/inbox/wwLab_automatisk_hamtning/
Dry run: process call will not be executed!!
```

**Review by Mropat — APPROVED** [review]

**Comment by Mropat** [comment]
```
Building wheels for collected packages: cg
  Building wheel for cg (PEP 517) ... done
  Created wheel for cg: filename=cg-20.21.0-py3-none-any.whl size=433346 sha256=74645dacd87edf0c7fc79ccf98e68be4c332b70e20383bfb12beb8c76360a3a1
  Stored in directory: /home/maryia.ropart/tmp/pip-ephem-wheel-cache-w_nknhjc/wheels/8c/1e/b8/816ed4facf8c5628b55966d3f2448eeb6ca70f35deff59389e
Successfully built cg
Installing collected packages: cg
  Attempting uninstall: cg
    Found existing installation: cg 20.20.1
    Uninstalling cg-20.20.1:
      Successfully uninstalled cg-20.20.1
Successfully installed cg-20.21.0

```

**Pull request merged**